# Group Project 604

> Group Member:

> Chang Ting Yu $\bigstar$G2

> Seah Yan De Bryan $\bigstar$G1

> Wang Li $\bigstar$G1

> Lyu Jiaming $\bigstar$G1

In [1]:
import numpy as np
import datetime
import pandas as pd
import matplotlib.pyplot as plt
import statsmodels.formula.api as smf
import statsmodels.api as sm
from finance_byu.regtables import Regtable
import warnings
warnings.filterwarnings('ignore')
from patsy.builtins import *

In [2]:
df1 = pd.read_csv('GPEX1set1.csv')
df2 = pd.read_csv('GPEX1set2.csv')
df3 = pd.read_csv('GPEX1set3.csv')

In [3]:
df1['Date'] = pd.to_datetime(df1.Date, format='%Y-%m-%d') 
df2['Date'] = pd.to_datetime(df2.Date, format='%Y-%m-%d') 
df3['Date'] = pd.to_datetime(df3.Date, format='%Y-%m-%d')

In [4]:
# convert the percentage return into decimal 
df = [df1,df2,df3]
for i in df:
    i['Return'] = i['Return'].apply(lambda x: x*0.01)

# Question 1

### Model_1

In [5]:
df1.head()

GVKEY       Date    Return  LogSize_-1  LogB/M_-1  Return_-2,-12
0   1000 1972-04-30  0.266667    2.819480  -0.266121      -0.461539
1   1000 1972-05-31 -0.070175    3.053069  -0.934455      -0.476744
2   1000 1972-06-30 -0.094340    2.977503  -0.861695      -0.173913
3   1000 1972-07-31  0.000000    2.875596  -0.762604      -0.196970
4   1000 1972-08-31 -0.020833    2.875935  -0.762604      -0.020408

In [6]:
formula_1 = 'Return ~ 1 + Q("LogSize_-1") + Q("LogB/M_-1") + Q("Return_-2,-12")'

def ols_coef(x,formula):
    return smf.ols(formula,data=x).fit()

reg_1 = df1.groupby('Date').apply(ols_coef, formula_1)
reg_1

Date
1964-04-30    <statsmodels.regression.linear_model.Regressio...
1964-05-31    <statsmodels.regression.linear_model.Regressio...
1964-06-30    <statsmodels.regression.linear_model.Regressio...
1964-07-31    <statsmodels.regression.linear_model.Regressio...
1964-08-31    <statsmodels.regression.linear_model.Regressio...
                                    ...                        
2020-11-30    <statsmodels.regression.linear_model.Regressio...
2020-12-31    <statsmodels.regression.linear_model.Regressio...
2021-01-31    <statsmodels.regression.linear_model.Regressio...
2021-02-28    <statsmodels.regression.linear_model.Regressio...
2021-03-31    <statsmodels.regression.linear_model.Regressio...
Length: 684, dtype: object

In [7]:
reg_1_list = list(reg_1.values)
model_1 = Regtable(reg_1_list,bfmt='.5f')
model_1 = model_1.render().T
model_1.index = reg_1.index
model_1 = model_1[model_1['obs']>=30]
model_1.drop(columns=[''], inplace =True)
model_1.rename(columns = lambda x: x.strip('Q("")'), inplace = True)
def func(model):
    for i in model.columns:
        model[i] = model[i].apply(lambda x: pd.to_numeric(x))
    return model
model_1 = func(model_1)
model_1

Intercept  LogSize_-1  LogB/M_-1  Return_-2,-12   obs   Rsq
Date                                                                   
1967-04-30   -0.00738    -0.00044    0.00484        0.12707    39  0.32
1967-05-31   -0.03699     0.00153   -0.00786       -0.14648    41  0.23
1967-06-30    0.16817    -0.02709   -0.00210        0.02727    42  0.12
1967-07-31    0.23555    -0.02547    0.03809        0.06306    42  0.19
1967-08-31    0.07589    -0.01062    0.03229       -0.01055    42  0.08
...               ...         ...        ...            ...   ...   ...
2020-11-30    0.28855    -0.01202    0.01752       -0.04559  3500  0.03
2020-12-31    0.16384    -0.01281   -0.01029       -0.01545  3468  0.01
2021-01-31    0.22090    -0.01622    0.06725        0.29714  1173  0.04
2021-02-28    0.19433    -0.00798    0.03927        0.01939  1004  0.06
2021-03-31    0.07015     0.00147    0.02026       -0.01333   923  0.05

[648 rows x 6 columns]

### Model_2

In [8]:
df2.head()

GVKEY       Date    Return  LogSize_-1  LogB/M_-1  Return_-2,-12  \
0   1000 1974-04-30  0.095238    1.924431  -0.081561      -0.428572   
1   1000 1974-05-31 -0.130435    2.016424   0.048045      -0.222223   
2   1000 1974-06-30  0.050000    1.877682   0.187806      -0.080001   
3   1000 1974-07-31 -0.142857    1.927492   0.139016      -0.166668   
4   1000 1974-08-31  0.166667    1.766826   0.293167      -0.322582   

   LogIssues_-1,-36  Accruals_Yr-1  ROA_Yr-1  LogAG_Yr-1  
0         -0.298542         -0.246    0.0894    0.089507  
1         -0.297521         -0.246    0.0894    0.089507  
2         -0.296501         -0.246    0.0894    0.089507  
3         -0.295198         -0.246    0.0894    0.089507  
4         -0.301428         -0.246    0.0894    0.089507

In [9]:
formula_2 = 'Return ~ 1 + Q("LogSize_-1") + Q("LogB/M_-1") + Q("Return_-2,-12") + Q("LogIssues_-1,-36") + Q("Accruals_Yr-1") + Q("ROA_Yr-1")+Q("LogAG_Yr-1")'

def ols_coef(x,formula):
    return smf.ols(formula,data=x).fit()

reg_2 = df2.groupby('Date').apply(ols_coef, formula_2)
reg_2

Date
1966-04-30    <statsmodels.regression.linear_model.Regressio...
1966-05-31    <statsmodels.regression.linear_model.Regressio...
1966-06-30    <statsmodels.regression.linear_model.Regressio...
1966-07-31    <statsmodels.regression.linear_model.Regressio...
1966-08-31    <statsmodels.regression.linear_model.Regressio...
                                    ...                        
2020-11-30    <statsmodels.regression.linear_model.Regressio...
2020-12-31    <statsmodels.regression.linear_model.Regressio...
2021-01-31    <statsmodels.regression.linear_model.Regressio...
2021-02-28    <statsmodels.regression.linear_model.Regressio...
2021-03-31    <statsmodels.regression.linear_model.Regressio...
Length: 660, dtype: object

In [10]:
reg_2_list = list(reg_2.values)
model_2 = Regtable(reg_2_list,bfmt='.5f')
model_2 = model_2.render().T
model_2.index = reg_2.index
model_2 = model_2[model_2['obs']>=30]
model_2.drop(columns=[''], inplace =True)
model_2.rename(columns = lambda x: x.strip('Q("")'), inplace = True)
def func(model):
    for i in model.columns:
        model[i] = model[i].apply(lambda x: pd.to_numeric(x))
    return model
model_2 = func(model_2)
model_2

Intercept  LogSize_-1  LogB/M_-1  Return_-2,-12  LogIssues_-1,-36  \
Date                                                                            
1969-09-30   -0.04868    -0.00815   -0.06929        0.01889           0.01162   
1969-10-31    0.09873    -0.01920   -0.07038        0.02837           0.01117   
1969-11-30    0.00329    -0.00581    0.01364        0.00151          -0.06569   
1969-12-31   -0.11957     0.02673    0.02111        0.23254          -0.00849   
1970-01-31   -0.03814     0.00980    0.02425        0.02928           0.00941   
...               ...         ...        ...            ...               ...   
2020-11-30    0.25473    -0.00520    0.03873       -0.06232          -0.00323   
2020-12-31    0.18357    -0.01518   -0.00908       -0.01546          -0.01158   
2021-01-31    0.19686     0.00457    0.11289        0.34864          -0.46666   
2021-02-28    0.36595    -0.03160    0.02774       -0.00318          -0.02741   
2021-03-31    0.00546     0.00703    0.02218        0.00564          -0.01147   

            Accruals_Yr-1  ROA_Yr-1  LogAG_Yr-1   obs   Rsq  
Date                                                         
1969-09-30       -0.00075   0.60453     0.04038    32  0.21  
1969-10-31       -0.00114  -0.31372     0.04194    33  0.16  
1969-11-30        0.00002   0.07845    -0.04405    33  0.18  
1969-12-31        0.00063   0.17656    -0.10238    35  0.41  
1970-01-31        0.00081  -1.02941    -0.03723    36  0.10  
...                   ...       ...         ...   ...   ...  
2020-11-30       -0.00001  -0.16760     0.01483  2071  0.07  
2020-12-31       -0.00000   0.00739    -0.02707  2068  0.02  
2021-01-31        0.00008  -0.64401    -0.18056   644  0.08  
2021-02-28       -0.00004   0.11076     0.05909   504  0.10  
2021-03-31       -0.00000   0.13676     0.04263   436  0.06  

[619 rows x 10 columns]

#### Model_3

In [11]:
df3.head()

GVKEY       Date    Return  LogSize_-1  LogB/M_-1  Return_-2,-12  \
0   1000 1974-04-30  0.095238    1.924431  -0.081561      -0.428572   
1   1000 1974-05-31 -0.130435    2.016424   0.048045      -0.222223   
2   1000 1974-06-30  0.050000    1.877682   0.187806      -0.080001   
3   1000 1974-07-31 -0.142857    1.927492   0.139016      -0.166668   
4   1000 1974-08-31  0.166667    1.766826   0.293167      -0.322582   

   LogIssues_-1,-36  Accruals_Yr-1  ROA_Yr-1  LogAG_Yr-1  DY_-1,-12  \
0         -0.298542         -0.246    0.0894    0.089507        0.0   
1         -0.297521         -0.246    0.0894    0.089507        0.0   
2         -0.296501         -0.246    0.0894    0.089507        0.0   
3         -0.295198         -0.246    0.0894    0.089507        0.0   
4         -0.301428         -0.246    0.0894    0.089507        0.0   

   LogReturn_-13,-36  LogIssues_-1,-12  Turnover_-1,-12  Debt/Price_Yr-1  \
0          -0.551282         -0.153494      9739.004887         1.021711   
1          -0.686046         -0.150169      9758.016991         0.931915   
2          -0.637681         -0.146841      9405.625892         1.070609   
3          -0.636363         -0.132863      9020.064092         1.018589   
4          -0.367346         -0.126248      8754.809368         1.196121   

   Sales/Price_Yr-1  
0          5.509943  
1          5.025683  
2          5.773643  
3          5.493106  
4          6.450510

In [12]:
formula_3 = 'Return ~ 1 + Q("LogSize_-1") + Q("LogB/M_-1") + Q("Return_-2,-12") + Q("LogIssues_-1,-36") + Q("Accruals_Yr-1") + Q("ROA_Yr-1") + Q("LogAG_Yr-1") + Q("DY_-1,-12") + Q("LogReturn_-13,-36") + Q("LogIssues_-1,-12") + Q("LogIssues_-1,-12") + Q("Turnover_-1,-12") + Q("Debt/Price_Yr-1") + Q("Sales/Price_Yr-1")'
def ols_coef(x,formula):
    return smf.ols(formula,data=x).fit()

reg_3 = df3.groupby('Date').apply(ols_coef, formula_3)
reg_3

Date
1968-03-31    <statsmodels.regression.linear_model.Regressio...
1968-04-30    <statsmodels.regression.linear_model.Regressio...
1968-05-31    <statsmodels.regression.linear_model.Regressio...
1968-06-30    <statsmodels.regression.linear_model.Regressio...
1968-07-31    <statsmodels.regression.linear_model.Regressio...
                                    ...                        
2020-11-30    <statsmodels.regression.linear_model.Regressio...
2020-12-31    <statsmodels.regression.linear_model.Regressio...
2021-01-31    <statsmodels.regression.linear_model.Regressio...
2021-02-28    <statsmodels.regression.linear_model.Regressio...
2021-03-31    <statsmodels.regression.linear_model.Regressio...
Length: 637, dtype: object

In [13]:
reg_3_list = list(reg_3.values)
model_3 = Regtable(reg_3_list,bfmt='.5f')
model_3 = model_3.render().T
model_3.index = reg_3.index
model_3 = model_3[model_3['obs']>=30]
model_3.drop(columns=[''], inplace =True)
model_3.rename(columns = lambda x: x.strip('Q("")'), inplace = True)
def func(model):
    for i in model.columns:
        model[i] = model[i].apply(lambda x: pd.to_numeric(x))
    return model
model_3 = func(model_3)
model_3

Intercept  LogSize_-1  LogB/M_-1  Return_-2,-12  LogIssues_-1,-36  \
Date                                                                            
1970-01-31   -0.00176    -0.01550    0.00973       -0.16299           0.04074   
1970-02-28   -0.15158     0.00724   -0.04496       -0.08526           0.04287   
1970-03-31    0.03957     0.00088    0.04793       -0.01908          -0.02577   
1970-04-30   -0.22475     0.02015    0.03795       -0.02076           0.00334   
1970-05-31   -0.10468     0.00615    0.00533       -0.05297          -0.03959   
...               ...         ...        ...            ...               ...   
2020-11-30    0.20875    -0.01043    0.01692       -0.06548           0.04458   
2020-12-31    0.18833    -0.01437   -0.00384       -0.00257          -0.03491   
2021-01-31    0.17460    -0.02715   -0.00191        0.06307           0.02356   
2021-02-28    0.13817    -0.00488    0.02378       -0.02481           0.00280   
2021-03-31   -0.05429     0.00858    0.00873        0.01805          -0.01064   

            Accruals_Yr-1  ROA_Yr-1  LogAG_Yr-1  DY_-1,-12  LogReturn_-13,-36  \
Date                                                                            
1970-01-31       -0.00030  -1.67905     0.06983    1.76905            0.05018   
1970-02-28       -0.00155  -0.95833     0.03589    0.70511            0.06894   
1970-03-31        0.00027   0.33329     0.01673   -0.40009            0.00506   
1970-04-30        0.00009   0.15797    -0.07797    1.11820           -0.00164   
1970-05-31       -0.00079   0.30006     0.00579   -0.20718           -0.00187   
...                   ...       ...         ...        ...                ...   
2020-11-30       -0.00000  -0.04107    -0.02445   -0.19405            0.00685   
2020-12-31       -0.00001  -0.02154    -0.03945   -0.63973            0.01301   
2021-01-31       -0.00002  -0.01022    -0.03090    0.18800           -0.01516   
2021-02-28        0.00000   0.06006    -0.01892   -0.70048           -0.01488   
2021-03-31        0.00001   0.11356     0.03513    0.76894            0.01983   

            LogIssues_-1,-12  Turnover_-1,-12  Debt/Price_Yr-1  \
Date                                                             
1970-01-31          -0.04376             -0.0          0.01640   
1970-02-28          -0.45283              0.0          0.00331   
1970-03-31          -0.04150             -0.0         -0.00588   
1970-04-30           0.02444             -0.0          0.00414   
1970-05-31           0.08803             -0.0         -0.02384   
...                      ...              ...              ...   
2020-11-30          -0.09933              0.0          0.02849   
2020-12-31           0.08128              0.0          0.00242   
2021-01-31           0.02413              0.0         -0.01350   
2021-02-28          -0.05718              0.0          0.03009   
2021-03-31          -0.05090             -0.0          0.05031   

            Sales/Price_Yr-1   obs   Rsq  
Date                                      
1970-01-31          -0.04346    30  0.68  
1970-02-28          -0.00253    32  0.61  
1970-03-31          -0.00691    60  0.09  
1970-04-30          -0.00710    89  0.50  
1970-05-31           0.00706    87  0.13  
...                      ...   ...   ...  
2020-11-30           0.00593  1816  0.10  
2020-12-31          -0.00162  1833  0.04  
2021-01-31           0.00166   589  0.24  
2021-02-28           0.00460   471  0.16  
2021-03-31           0.00356   412  0.18  

[615 rows x 16 columns]

# Question 2

### Model_1

In [14]:
def model_summary(model):
    s = model.describe().T
    s['std_error'] = s['std']/np.sqrt(s['count'])
    s['tstat'] = s['mean']/s['std_error']
    return s[['mean','tstat']].iloc[:-2]

In [15]:
model_1_summary = model_summary(model_1)
model_1_summary

mean     tstat
Intercept      0.017794  5.582529
LogSize_-1    -0.001086 -2.930749
LogB/M_-1      0.004723  7.565304
Return_-2,-12  0.009976  6.626130

In [16]:
def add_asterisks(mean, t_statistic):
    if abs(t_statistic) >= 2.5834:
        return '{:.6f}{}'.format(mean, '***')
    elif abs(t_statistic) >= 1.9636:
        return '{:.6f}{}'.format(mean, '**')
    elif abs(t_statistic) >= 1.6472:
        return '{:.6f}{}'.format(mean, '*')
    else:
        return '{:.6f}'.format(mean)

In [17]:
model_1_summary['mean'] = model_1_summary.apply(lambda row: add_asterisks(row['mean'], row['tstat']), axis=1)
model_1_summary_T = model_1_summary.T
model_1_summary_T

Intercept    LogSize_-1    LogB/M_-1 Return_-2,-12
mean   0.017794***  -0.001086***  0.004723***   0.009976***
tstat     5.582529     -2.930749     7.565304       6.62613

In [18]:
def data_transform(model):
    M = pd.DataFrame(data = None)
    for i in model.columns:
        a = pd.DataFrame(model[i])
        a = a.rename(columns = {i:'Model_1'})
        a.index = ['Average of time series of ' + i,''] # set tstat as empty as required 
        M = M.append(a)
    return M
model_1_final = data_transform(model_1_summary_T)
model_1_final

Model_1
Average of time series of Intercept       0.017794***
                                             5.582529
Average of time series of LogSize_-1     -0.001086***
                                            -2.930749
Average of time series of LogB/M_-1       0.004723***
                                             7.565304
Average of time series of Return_-2,-12   0.009976***
                                              6.62613

### Model_2

In [19]:
model_2_summary = model_summary(model_2)
model_2_summary

mean     tstat
Intercept         0.017962  5.608902
LogSize_-1       -0.001019 -2.887870
LogB/M_-1         0.003916  6.236442
Return_-2,-12     0.008599  5.820194
LogIssues_-1,-36 -0.001977 -1.795763
Accruals_Yr-1    -0.000008 -1.616272
ROA_Yr-1          0.016454  3.140100
LogAG_Yr-1       -0.007169 -5.266588

In [20]:
# mark the tstat with * for different significance level degree of freedom is 619-1
def add_asterisks(mean, t_statistic):
    if abs(t_statistic) >= 2.5834:
        return '{:.6f}{}'.format(mean, '***')
    elif abs(t_statistic) >= 1.9636:
        return '{:.6f}{}'.format(mean, '**')
    elif abs(t_statistic) >= 1.6472:
        return '{:.6f}{}'.format(mean, '*')
    else:
        return '{:.6f}'.format(mean)
    
model_2_summary['mean'] = model_2_summary.apply(lambda row: add_asterisks(row['mean'], row['tstat']), axis=1)
model_2_summary_T = model_2_summary.T
model_2_summary_T

Intercept    LogSize_-1    LogB/M_-1 Return_-2,-12 LogIssues_-1,-36  \
mean   0.017962***  -0.001019***  0.003916***   0.008599***       -0.001977*   
tstat     5.608902      -2.88787     6.236442      5.820194        -1.795763   

      Accruals_Yr-1     ROA_Yr-1    LogAG_Yr-1  
mean      -0.000008  0.016454***  -0.007169***  
tstat     -1.616272       3.1401     -5.266588

In [21]:
def data_transform(model):
    M = pd.DataFrame(data = None)
    for i in model.columns:
        a = pd.DataFrame(model[i])
        a = a.rename(columns = {i:'Model_2'})
        a.index = ['Average of time series of ' + i,''] # set tstat as empty as required 
        M = M.append(a)
    return M
model_2_final = data_transform(model_2_summary_T)
model_2_final

Model_2
Average of time series of Intercept          0.017962***
                                                5.608902
Average of time series of LogSize_-1        -0.001019***
                                                -2.88787
Average of time series of LogB/M_-1          0.003916***
                                                6.236442
Average of time series of Return_-2,-12      0.008599***
                                                5.820194
Average of time series of LogIssues_-1,-36    -0.001977*
                                               -1.795763
Average of time series of Accruals_Yr-1        -0.000008
                                               -1.616272
Average of time series of ROA_Yr-1           0.016454***
                                                  3.1401
Average of time series of LogAG_Yr-1        -0.007169***
                                               -5.266588

### Model 3

In [22]:
model_3_summary = model_summary(model_3)
model_3_summary

mean     tstat
Intercept          0.016642  5.690490
LogSize_-1        -0.000866 -2.666004
LogB/M_-1          0.002928  5.934736
Return_-2,-12      0.007302  5.684287
LogIssues_-1,-36  -0.001240 -1.336702
Accruals_Yr-1     -0.000008 -1.796735
ROA_Yr-1           0.014103  2.662801
LogAG_Yr-1        -0.005822 -4.898245
DY_-1,-12         -0.014077 -0.839710
LogReturn_-13,-36 -0.000570 -1.030280
LogIssues_-1,-12   0.002840  1.315398
Turnover_-1,-12    0.000000       NaN
Debt/Price_Yr-1   -0.000595 -1.539125
Sales/Price_Yr-1   0.000406  2.749419

In [23]:
# mark the tstat with * for different significance level degree of freedom is 615-1
def add_asterisks(mean, t_statistic):
    if abs(t_statistic) >= 2.5839:
        return '{:.6f}{}'.format(mean, '***')
    elif abs(t_statistic) >= 1.9638:
        return '{:.6f}{}'.format(mean, '**')
    elif abs(t_statistic) >= 1.6473:
        return '{:.6f}{}'.format(mean, '*')
    else:
        return '{:.6f}'.format(mean)
model_3_summary['mean'] = model_3_summary.apply(lambda row: add_asterisks(row['mean'], row['tstat']), axis=1)
model_3_summary_T = model_3_summary.T
model_3_summary_T

Intercept    LogSize_-1    LogB/M_-1 Return_-2,-12 LogIssues_-1,-36  \
mean   0.016642***  -0.000866***  0.002928***   0.007302***        -0.001240   
tstat      5.69049     -2.666004     5.934736      5.684287        -1.336702   

      Accruals_Yr-1     ROA_Yr-1    LogAG_Yr-1  DY_-1,-12 LogReturn_-13,-36  \
mean     -0.000008*  0.014103***  -0.005822***  -0.014077         -0.000570   
tstat     -1.796735     2.662801     -4.898245   -0.83971          -1.03028   

      LogIssues_-1,-12 Turnover_-1,-12 Debt/Price_Yr-1 Sales/Price_Yr-1  
mean          0.002840        0.000000       -0.000595      0.000406***  
tstat         1.315398             NaN       -1.539125         2.749419

In [24]:
def data_transform(model):
    M = pd.DataFrame(data = None)
    for i in model.columns:
        a = pd.DataFrame(model[i])
        a = a.rename(columns = {i:'Model_3'})
        a.index = ['Average of time series of ' + i,''] # set tstat as empty as required 
        M = M.append(a)
    return M
model_3_final = data_transform(model_3_summary_T)
model_3_final

Model_3
Average of time series of Intercept           0.016642***
                                                  5.69049
Average of time series of LogSize_-1         -0.000866***
                                                -2.666004
Average of time series of LogB/M_-1           0.002928***
                                                 5.934736
Average of time series of Return_-2,-12       0.007302***
                                                 5.684287
Average of time series of LogIssues_-1,-36      -0.001240
                                                -1.336702
Average of time series of Accruals_Yr-1        -0.000008*
                                                -1.796735
Average of time series of ROA_Yr-1            0.014103***
                                                 2.662801
Average of time series of LogAG_Yr-1         -0.005822***
                                                -4.898245
Average of time series of DY_-1,-12             -0.014077
                                                 -0.83971
Average of time series of LogReturn_-13,-36     -0.000570
                                                 -1.03028
Average of time series of LogIssues_-1,-12       0.002840
                                                 1.315398
Average of time series of Turnover_-1,-12        0.000000
                                                      NaN
Average of time series of Debt/Price_Yr-1       -0.000595
                                                -1.539125
Average of time series of Sales/Price_Yr-1    0.000406***
                                                 2.749419

In [25]:
combined_model = pd.concat([model_1_final,model_2_final,model_3_final], axis = 0)
combined_model = combined_model.fillna('')
combined_model

Model_1       Model_2  \
Average of time series of Intercept           0.017794***                 
                                                 5.582529                 
Average of time series of LogSize_-1         -0.001086***                 
                                                -2.930749                 
Average of time series of LogB/M_-1           0.004723***                 
                                                 7.565304                 
Average of time series of Return_-2,-12       0.009976***                 
                                                  6.62613                 
Average of time series of Intercept                         0.017962***   
                                                               5.608902   
Average of time series of LogSize_-1                       -0.001019***   
                                                               -2.88787   
Average of time series of LogB/M_-1                         0.003916***   
                                                               6.236442   
Average of time series of Return_-2,-12                     0.008599***   
                                                               5.820194   
Average of time series of LogIssues_-1,-36                   -0.001977*   
                                                              -1.795763   
Average of time series of Accruals_Yr-1                       -0.000008   
                                                              -1.616272   
Average of time series of ROA_Yr-1                          0.016454***   
                                                                 3.1401   
Average of time series of LogAG_Yr-1                       -0.007169***   
                                                              -5.266588   
Average of time series of Intercept                                       
                                                                          
Average of time series of LogSize_-1                                      
                                                                          
Average of time series of LogB/M_-1                                       
                                                                          
Average of time series of Return_-2,-12                                   
                                                                          
Average of time series of LogIssues_-1,-36                                
                                                                          
Average of time series of Accruals_Yr-1                                   
                                                                          
Average of time series of ROA_Yr-1                                        
                                                                          
Average of time series of LogAG_Yr-1                                      
                                                                          
Average of time series of DY_-1,-12                                       
                                                                          
Average of time series of LogReturn_-13,-36                               
                                                                          
Average of time series of LogIssues_-1,-12                                
                                                                          
Average of time series of Turnover_-1,-12                                 
                                                                          
Average of time series of Debt/Price_Yr-1                                 
                                                                          
Average of time series of Sales/Price_Yr-1                                
                                                                          

                                                  Model_3  
Average of tim

# Question 3

In [26]:
model_3.head()

Intercept  LogSize_-1  LogB/M_-1  Return_-2,-12  LogIssues_-1,-36  \
Date                                                                            
1970-01-31   -0.00176    -0.01550    0.00973       -0.16299           0.04074   
1970-02-28   -0.15158     0.00724   -0.04496       -0.08526           0.04287   
1970-03-31    0.03957     0.00088    0.04793       -0.01908          -0.02577   
1970-04-30   -0.22475     0.02015    0.03795       -0.02076           0.00334   
1970-05-31   -0.10468     0.00615    0.00533       -0.05297          -0.03959   

            Accruals_Yr-1  ROA_Yr-1  LogAG_Yr-1  DY_-1,-12  LogReturn_-13,-36  \
Date                                                                            
1970-01-31       -0.00030  -1.67905     0.06983    1.76905            0.05018   
1970-02-28       -0.00155  -0.95833     0.03589    0.70511            0.06894   
1970-03-31        0.00027   0.33329     0.01673   -0.40009            0.00506   
1970-04-30        0.00009   0.15797    -0.07797    1.11820           -0.00164   
1970-05-31       -0.00079   0.30006     0.00579   -0.20718           -0.00187   

            LogIssues_-1,-12  Turnover_-1,-12  Debt/Price_Yr-1  \
Date                                                             
1970-01-31          -0.04376             -0.0          0.01640   
1970-02-28          -0.45283              0.0          0.00331   
1970-03-31          -0.04150             -0.0         -0.00588   
1970-04-30           0.02444             -0.0          0.00414   
1970-05-31           0.08803             -0.0         -0.02384   

            Sales/Price_Yr-1  obs   Rsq  
Date                                     
1970-01-31          -0.04346   30  0.68  
1970-02-28          -0.00253   32  0.61  
1970-03-31          -0.00691   60  0.09  
1970-04-30          -0.00710   89  0.50  
1970-05-31           0.00706   87  0.13

In [27]:
df3.head()

GVKEY       Date    Return  LogSize_-1  LogB/M_-1  Return_-2,-12  \
0   1000 1974-04-30  0.095238    1.924431  -0.081561      -0.428572   
1   1000 1974-05-31 -0.130435    2.016424   0.048045      -0.222223   
2   1000 1974-06-30  0.050000    1.877682   0.187806      -0.080001   
3   1000 1974-07-31 -0.142857    1.927492   0.139016      -0.166668   
4   1000 1974-08-31  0.166667    1.766826   0.293167      -0.322582   

   LogIssues_-1,-36  Accruals_Yr-1  ROA_Yr-1  LogAG_Yr-1  DY_-1,-12  \
0         -0.298542         -0.246    0.0894    0.089507        0.0   
1         -0.297521         -0.246    0.0894    0.089507        0.0   
2         -0.296501         -0.246    0.0894    0.089507        0.0   
3         -0.295198         -0.246    0.0894    0.089507        0.0   
4         -0.301428         -0.246    0.0894    0.089507        0.0   

   LogReturn_-13,-36  LogIssues_-1,-12  Turnover_-1,-12  Debt/Price_Yr-1  \
0          -0.551282         -0.153494      9739.004887         1.021711   
1          -0.686046         -0.150169      9758.016991         0.931915   
2          -0.637681         -0.146841      9405.625892         1.070609   
3          -0.636363         -0.132863      9020.064092         1.018589   
4          -0.367346         -0.126248      8754.809368         1.196121   

   Sales/Price_Yr-1  
0          5.509943  
1          5.025683  
2          5.773643  
3          5.493106  
4          6.450510

In [28]:
moving_average = pd. DataFrame()
moving_average["Intercept_120"] = model_3["Intercept"].rolling(window = 120).mean()
moving_average["LogSize_-1_120"] = model_3["LogSize_-1"].rolling(window = 120).mean()
moving_average["LogB/M_-1_120"] = model_3["LogB/M_-1"].rolling(window = 120).mean()
moving_average["Return_-2,-12_120"] = model_3["Return_-2,-12"].rolling(window = 120).mean()
moving_average["LogIssues_-1,-36_120"] = model_3["LogIssues_-1,-36"].rolling(window = 120).mean()
moving_average["Accruals_Yr-1_120"] = model_3["Accruals_Yr-1"].rolling(window = 120).mean()
moving_average["ROA_Yr-1_120"] = model_3["ROA_Yr-1"].rolling(window = 120).mean()
moving_average["LogAG_Yr-1_120"] = model_3["LogAG_Yr-1"].rolling(window = 120).mean()
moving_average["DY_-1,-12_120"] = model_3["DY_-1,-12"].rolling(window = 120).mean()
moving_average["LogReturn_-13,-36_120"] = model_3["LogReturn_-13,-36"].rolling(window = 120).mean()
moving_average["LogIssues_-1,-12_120"] = model_3["LogIssues_-1,-12"].rolling(window = 120).mean()
moving_average["Turnover_-1,-12_120"] = model_3["Turnover_-1,-12"].rolling(window = 120).mean()
moving_average["Debt/Price_Yr-1_120"] = model_3["Debt/Price_Yr-1"].rolling(window = 120).mean()
moving_average["Sales/Price_Yr-1_120"] = model_3["Sales/Price_Yr-1"].rolling(window = 120).mean()
moving_average = moving_average.dropna()
moving_average

Intercept_120  LogSize_-1_120  LogB/M_-1_120  Return_-2,-12_120  \
Date                                                                          
1979-12-31       0.011610       -0.000828       0.006093           0.011138   
1980-01-31       0.013897       -0.000838       0.006087           0.012823   
1980-02-29       0.014803       -0.000878       0.006458           0.013802   
1980-03-31       0.012611       -0.000786       0.006093           0.013711   
1980-04-30       0.014685       -0.000992       0.005698           0.014181   
...                   ...             ...            ...                ...   
2020-11-30       0.012733       -0.000350      -0.001768           0.001887   
2020-12-31       0.013630       -0.000474      -0.001818           0.002069   
2021-01-31       0.014580       -0.000644      -0.001858           0.002717   
2021-02-28       0.015326       -0.000683      -0.001722           0.002265   
2021-03-31       0.014730       -0.000615      -0.001550           0.002381   

            LogIssues_-1,-36_120  Accruals_Yr-1_120  ROA_Yr-1_120  \
Date                                                                
1979-12-31             -0.001399      -3.516667e-05      0.025374   
1980-01-31             -0.001846      -3.341667e-05      0.035942   
1980-02-29             -0.002171      -2.141667e-05      0.043591   
1980-03-31             -0.002029      -2.325000e-05      0.041923   
1980-04-30             -0.002085      -2.400000e-05      0.041108   
...                          ...                ...           ...   
2020-11-30             -0.003778       7.500000e-07      0.000121   
2020-12-31             -0.004399       6.666667e-07      0.000391   
2021-01-31             -0.004353       6.666667e-07      0.000080   
2021-02-28             -0.004335       5.833333e-07      0.001151   
2021-03-31             -0.004376       5.833333e-07      0.001966   

            LogAG_Yr-1_120  DY_-1,-12_120  LogReturn_-13,-36_120  \
Date                                                               
1979-12-31       -0.000088      -0.016041              -0.000151   
1980-01-31       -0.000567      -0.040051              -0.000646   
1980-02-29       -0.001195      -0.048003              -0.001178   
1980-03-31       -0.001181      -0.037648              -0.001219   
1980-04-30       -0.000520      -0.042648              -0.001145   
...                    ...            ...                    ...   
2020-11-30       -0.004398      -0.057238              -0.000275   
2020-12-31       -0.004464      -0.059250               0.000034   
2021-01-31       -0.004310      -0.056270               0.000054   
2021-02-28       -0.004584      -0.058380              -0.000021   
2021-03-31       -0.004055      -0.051294               0.000092   

            LogIssues_-1,-12_120  Turnover_-1,-12_120  Debt/Price_Yr-1_120  \
Date                                                                         
1979-12-31             -0.005884                  0.0             0.002233   
1980-01-31             -0.005210                 -0.0             0.002047   
1980-02-29             -0.001473                  0.0             0.002043   
1980-03-31             -0.000991                 -0.0             0.002077   
1980-04-30             -0.001105                  0.0             0.002101   
...                          ...                  ...                  ...   
2020-11-30              0.002328                  0.0            -0.001216   
2020-12-31              0.002982                  0.0            -0.001173   
2021-01-31              0.003338                  0.0            -0.001349   
2021-02-28              0.003096                  0.0            -0.001183   
2021-03-31              0.002577                 -0.0            -0.000690   

            Sales/Price_Yr-1_120  
Date                              
1979-12-31             -0.000502  
1980-01-31             -0.000185  
1980-02-29             -0.000160  
1980-03-31 

In [29]:
start_date = '1979-12-31'
after_start_date = df3["Date"] >= start_date
df3_actual = df3.loc[after_start_date]
df3_actual

GVKEY       Date    Return  LogSize_-1  LogB/M_-1  Return_-2,-12  \
92         1004 1979-12-31  0.268817    2.980251   0.181569       0.208865   
93         1004 1980-01-31  0.226695    3.340787  -0.056517       0.163113   
94         1004 1980-02-29  0.026087    3.424113  -0.030764       0.275932   
95         1004 1980-03-31 -0.296610    3.548208  -0.056517       0.585662   
96         1004 1980-04-30  0.009639    3.196506   0.295327       0.576707   
...         ...        ...       ...         ...        ...            ...   
1357950  317264 2020-08-31 -0.011710    6.076221   0.661244      -0.154099   
1357951  317264 2020-09-30 -0.050948    6.063467   0.673023      -0.178057   
1357952  317264 2020-10-31  0.023720    6.010200   0.725315      -0.185330   
1357953  317264 2020-11-30  0.332927    6.027104   0.701871      -0.356627   
1357954  317264 2020-12-31  0.115279    6.307898   0.414494      -0.428572   

         LogIssues_-1,-36  Accruals_Yr-1  ROA_Yr-1  LogAG_Yr-1  DY_-1,-12  \
92               0.313326          2.457  0.039490    0.215465   0.036559   
93               0.436316          2.457  0.039490    0.215465   0.028814   
94               0.545934          2.457  0.039490    0.215465   0.036522   
95               0.645355          2.457  0.039490    0.215465   0.029153   
96               0.646576         -1.103  0.049795    0.127634   0.041446   
...                   ...            ...       ...         ...        ...   
1357950         -0.077902       -115.174 -0.030312   -0.065911   0.000000   
1357951         -0.078937       -115.174 -0.030312   -0.065911   0.000000   
1357952         -0.079973       -115.174 -0.030312   -0.065911   0.000000   
1357953         -0.086573       -115.174 -0.030312   -0.065911   0.000000   
1357954         -0.093217       -115.174 -0.030312   -0.065911   0.000000   

         LogReturn_-13,-36  LogIssues_-1,-12  Turnover_-1,-12  \
92                0.126577          0.009093     18305.338137   
93                0.338005          0.131147     21231.256206   
94                0.825340          0.239829     30526.102419   
95                0.639495          0.337378     32507.391326   
96                0.797559          0.336727     33503.947702   
...                    ...               ...              ...   
1357950           0.270834         -0.071939    252262.069994   
1357951           0.514577         -0.070355    252022.588874   
1357952           0.519062         -0.067314    257688.600552   
1357953           0.741259         -0.069822    252706.685796   
1357954           0.987534         -0.072356    255335.436094   

         Debt/Price_Yr-1  Sales/Price_Yr-1  
92              1.361669          4.713410  
93              0.949495          3.286672  
94              0.873583          3.023904  
95              0.771632          2.671002  
96              1.282982          4.854060  
...                  ...               ...  
1357950         1.598806          0.368673  
1357951         1.619327          0.373405  
1357952         1.707922          0.393834  
1357953         1.679295          0.387233  
1357954         1.268176          0.292432  

[1206255 rows x 16 columns]

In [30]:
df3_predict = pd.merge(df3_actual, moving_average, how ='left', on ='Date')
df3_predict

GVKEY       Date    Return  LogSize_-1  LogB/M_-1  Return_-2,-12  \
0          1004 1979-12-31  0.268817    2.980251   0.181569       0.208865   
1          1004 1980-01-31  0.226695    3.340787  -0.056517       0.163113   
2          1004 1980-02-29  0.026087    3.424113  -0.030764       0.275932   
3          1004 1980-03-31 -0.296610    3.548208  -0.056517       0.585662   
4          1004 1980-04-30  0.009639    3.196506   0.295327       0.576707   
...         ...        ...       ...         ...        ...            ...   
1206250  317264 2020-08-31 -0.011710    6.076221   0.661244      -0.154099   
1206251  317264 2020-09-30 -0.050948    6.063467   0.673023      -0.178057   
1206252  317264 2020-10-31  0.023720    6.010200   0.725315      -0.185330   
1206253  317264 2020-11-30  0.332927    6.027104   0.701871      -0.356627   
1206254  317264 2020-12-31  0.115279    6.307898   0.414494      -0.428572   

         LogIssues_-1,-36  Accruals_Yr-1  ROA_Yr-1  LogAG_Yr-1  ...  \
0                0.313326          2.457  0.039490    0.215465  ...   
1                0.436316          2.457  0.039490    0.215465  ...   
2                0.545934          2.457  0.039490    0.215465  ...   
3                0.645355          2.457  0.039490    0.215465  ...   
4                0.646576         -1.103  0.049795    0.127634  ...   
...                   ...            ...       ...         ...  ...   
1206250         -0.077902       -115.174 -0.030312   -0.065911  ...   
1206251         -0.078937       -115.174 -0.030312   -0.065911  ...   
1206252         -0.079973       -115.174 -0.030312   -0.065911  ...   
1206253         -0.086573       -115.174 -0.030312   -0.065911  ...   
1206254         -0.093217       -115.174 -0.030312   -0.065911  ...   

         LogIssues_-1,-36_120  Accruals_Yr-1_120  ROA_Yr-1_120  \
0                   -0.001399      -3.516667e-05      0.025374   
1                   -0.001846      -3.341667e-05      0.035942   
2                   -0.002171      -2.141667e-05      0.043591   
3                   -0.002029      -2.325000e-05      0.041923   
4                   -0.002085      -2.400000e-05      0.041108   
...                       ...                ...           ...   
1206250             -0.003644       9.166667e-07     -0.000179   
1206251             -0.003820       8.333333e-07      0.000503   
1206252             -0.003940       8.333333e-07      0.000882   
1206253             -0.003778       7.500000e-07      0.000121   
1206254             -0.004399       6.666667e-07      0.000391   

         LogAG_Yr-1_120  DY_-1,-12_120  LogReturn_-13,-36_120  \
0             -0.000088      -0.016041              -0.000151   
1             -0.000567      -0.040051              -0.000646   
2             -0.001195      -0.048003              -0.001178   
3             -0.001181      -0.037648              -0.001219   
4             -0.000520      -0.042648              -0.001145   
...                 ...            ...                    ...   
1206250       -0.005128      -0.062232              -0.000360   
1206251       -0.004828      -0.059372              -0.000257   
1206252       -0.004506      -0.058341              -0.000211   
1206253       -0.004398      -0.057238              -0.000275   
1206254       -0.004464      -0.059250               0.000034   

         LogIssues_-1,-12_120  Turnover_-1,-12_120  Debt/Price_Yr-1_120  \
0                   -0.005884                  0.0             0.002233   
1                   -0.005210                 -0.0             0.002047   
2                   -0.001473                  0.0             0.002043   
3                   -0.000991                 -0.0             0.002077   
4                   -0.001105                  0.0             0.002101   
...                       ...                  ...                  ...   
1206250              0.003083                  0.0            -0.001372   
1206251              0.003518                 

In [31]:
df3_predict['Forecast/Predicted Return']\
= df3_predict['LogSize_-1']*df3_predict['LogSize_-1_120'] + df3_predict['LogB/M_-1']*df3_predict['LogB/M_-1_120']\
  + df3_predict['Return_-2,-12']*df3_predict['Return_-2,-12_120'] + df3_predict['LogIssues_-1,-36']*df3_predict['LogIssues_-1,-36_120']\
  + df3_predict['Accruals_Yr-1']*df3_predict['Accruals_Yr-1_120'] + df3_predict['ROA_Yr-1']*df3_predict['ROA_Yr-1_120']\
  + df3_predict['LogAG_Yr-1']*df3_predict['LogAG_Yr-1_120'] + df3_predict['DY_-1,-12']*df3_predict['DY_-1,-12_120']\
  + df3_predict['LogReturn_-13,-36']*df3_predict['LogReturn_-13,-36_120'] + df3_predict['LogIssues_-1,-12']*df3_predict['LogIssues_-1,-12_120']\
  + df3_predict['Turnover_-1,-12']*df3_predict['Turnover_-1,-12_120'] + df3_predict['Debt/Price_Yr-1']*df3_predict['Debt/Price_Yr-1_120']\
  + df3_predict['Sales/Price_Yr-1']*df3_predict['Sales/Price_Yr-1_120'] + df3_predict['Intercept_120']
df3_predict

GVKEY       Date    Return  LogSize_-1  LogB/M_-1  Return_-2,-12  \
0          1004 1979-12-31  0.268817    2.980251   0.181569       0.208865   
1          1004 1980-01-31  0.226695    3.340787  -0.056517       0.163113   
2          1004 1980-02-29  0.026087    3.424113  -0.030764       0.275932   
3          1004 1980-03-31 -0.296610    3.548208  -0.056517       0.585662   
4          1004 1980-04-30  0.009639    3.196506   0.295327       0.576707   
...         ...        ...       ...         ...        ...            ...   
1206250  317264 2020-08-31 -0.011710    6.076221   0.661244      -0.154099   
1206251  317264 2020-09-30 -0.050948    6.063467   0.673023      -0.178057   
1206252  317264 2020-10-31  0.023720    6.010200   0.725315      -0.185330   
1206253  317264 2020-11-30  0.332927    6.027104   0.701871      -0.356627   
1206254  317264 2020-12-31  0.115279    6.307898   0.414494      -0.428572   

         LogIssues_-1,-36  Accruals_Yr-1  ROA_Yr-1  LogAG_Yr-1  ...  \
0                0.313326          2.457  0.039490    0.215465  ...   
1                0.436316          2.457  0.039490    0.215465  ...   
2                0.545934          2.457  0.039490    0.215465  ...   
3                0.645355          2.457  0.039490    0.215465  ...   
4                0.646576         -1.103  0.049795    0.127634  ...   
...                   ...            ...       ...         ...  ...   
1206250         -0.077902       -115.174 -0.030312   -0.065911  ...   
1206251         -0.078937       -115.174 -0.030312   -0.065911  ...   
1206252         -0.079973       -115.174 -0.030312   -0.065911  ...   
1206253         -0.086573       -115.174 -0.030312   -0.065911  ...   
1206254         -0.093217       -115.174 -0.030312   -0.065911  ...   

         Accruals_Yr-1_120  ROA_Yr-1_120  LogAG_Yr-1_120  DY_-1,-12_120  \
0            -3.516667e-05      0.025374       -0.000088      -0.016041   
1            -3.341667e-05      0.035942       -0.000567      -0.040051   
2            -2.141667e-05      0.043591       -0.001195      -0.048003   
3            -2.325000e-05      0.041923       -0.001181      -0.037648   
4            -2.400000e-05      0.041108       -0.000520      -0.042648   
...                    ...           ...             ...            ...   
1206250       9.166667e-07     -0.000179       -0.005128      -0.062232   
1206251       8.333333e-07      0.000503       -0.004828      -0.059372   
1206252       8.333333e-07      0.000882       -0.004506      -0.058341   
1206253       7.500000e-07      0.000121       -0.004398      -0.057238   
1206254       6.666667e-07      0.000391       -0.004464      -0.059250   

         LogReturn_-13,-36_120  LogIssues_-1,-12_120  Turnover_-1,-12_120  \
0                    -0.000151             -0.005884                  0.0   
1                    -0.000646             -0.005210                 -0.0   
2                    -0.001178             -0.001473                  0.0   
3                    -0.001219             -0.000991                 -0.0   
4                    -0.001145             -0.001105                  0.0   
...                        ...                   ...                  ...   
1206250              -0.000360              0.003083                  0.0   
1206251              -0.000257              0.003518                 -0.0   
1206252              -0.000211              0.002626                  0.0   
1206253              -0.000275              0.002328                  0.0   
1206254               0.000034              0.002982                  0.0   

         Debt/Price_Yr-1_120  Sales/Price_Yr-1_120  Forecast/Predicted Return  
0                   0.002233             -0.000502                   0.013048  
1                   0.002047             -0.000185                   0.012534  
2                   0.002043             -0.000160                   0.013856  
3                   0.002077             -0.000094                   0.016

In [32]:
df3_predict_return = pd.DataFrame()
df3_predict_return['Date'] = df3_predict['Date'] + pd.offsets.MonthEnd(1)
df3_predict_return['GVKEY'] = df3_predict['GVKEY']
df3_predict_return['Forecast/Predicted Return'] = df3_predict['Forecast/Predicted Return']
df3_predict_return

Date   GVKEY  Forecast/Predicted Return
0       1980-01-31    1004                   0.013048
1       1980-02-29    1004                   0.012534
2       1980-03-31    1004                   0.013856
3       1980-04-30    1004                   0.016681
4       1980-05-31    1004                   0.021450
...            ...     ...                        ...
1206250 2020-09-30  317264                   0.007863
1206251 2020-10-31  317264                   0.006656
1206252 2020-11-30  317264                   0.005834
1206253 2020-12-31  317264                   0.007153
1206254 2021-01-31  317264                   0.008193

[1206255 rows x 3 columns]

In [33]:
df3_predict_return.sort_values(by = 'Date')

Date   GVKEY  Forecast/Predicted Return
0       1980-01-31    1004                   0.013048
148119  1980-01-31    3419                   0.014066
218016  1980-01-31    4545                   0.008322
453930  1980-01-31    8400                   0.006845
65637   1980-01-31    2066                   0.013256
...            ...     ...                        ...
1173286 2021-04-30  176670                   0.015529
418066  2021-04-30    7881                   0.003928
1105240 2021-04-30  133288                   0.013159
605973  2021-04-30   10839                   0.012722
910137  2021-04-30   28408                   0.014951

[1206255 rows x 3 columns]

In [34]:
start_date2 = '1980-01-31'
after_start_date2 = df3["Date"] >= start_date2
df3_actual_return = df3.loc[after_start_date2]
df3_actual_return

GVKEY       Date    Return  LogSize_-1  LogB/M_-1  Return_-2,-12  \
93         1004 1980-01-31  0.226695    3.340787  -0.056517       0.163113   
94         1004 1980-02-29  0.026087    3.424113  -0.030764       0.275932   
95         1004 1980-03-31 -0.296610    3.548208  -0.056517       0.585662   
96         1004 1980-04-30  0.009639    3.196506   0.295327       0.576707   
97         1004 1980-05-31 -0.012048    3.196647   0.187944       0.222463   
...         ...        ...       ...         ...        ...            ...   
1357950  317264 2020-08-31 -0.011710    6.076221   0.661244      -0.154099   
1357951  317264 2020-09-30 -0.050948    6.063467   0.673023      -0.178057   
1357952  317264 2020-10-31  0.023720    6.010200   0.725315      -0.185330   
1357953  317264 2020-11-30  0.332927    6.027104   0.701871      -0.356627   
1357954  317264 2020-12-31  0.115279    6.307898   0.414494      -0.428572   

         LogIssues_-1,-36  Accruals_Yr-1  ROA_Yr-1  LogAG_Yr-1  DY_-1,-12  \
93               0.436316          2.457  0.039490    0.215465   0.028814   
94               0.545934          2.457  0.039490    0.215465   0.036522   
95               0.645355          2.457  0.039490    0.215465   0.029153   
96               0.646576         -1.103  0.049795    0.127634   0.041446   
97               0.647798         -1.103  0.049795    0.127634   0.051084   
...                   ...            ...       ...         ...        ...   
1357950         -0.077902       -115.174 -0.030312   -0.065911   0.000000   
1357951         -0.078937       -115.174 -0.030312   -0.065911   0.000000   
1357952         -0.079973       -115.174 -0.030312   -0.065911   0.000000   
1357953         -0.086573       -115.174 -0.030312   -0.065911   0.000000   
1357954         -0.093217       -115.174 -0.030312   -0.065911   0.000000   

         LogReturn_-13,-36  LogIssues_-1,-12  Turnover_-1,-12  \
93                0.338005          0.131147     21231.256206   
94                0.825340          0.239829     30526.102419   
95                0.639495          0.337378     32507.391326   
96                0.797559          0.336727     33503.947702   
97                0.418066          0.336076     34249.060531   
...                    ...               ...              ...   
1357950           0.270834         -0.071939    252262.069994   
1357951           0.514577         -0.070355    252022.588874   
1357952           0.519062         -0.067314    257688.600552   
1357953           0.741259         -0.069822    252706.685796   
1357954           0.987534         -0.072356    255335.436094   

         Debt/Price_Yr-1  Sales/Price_Yr-1  
93              0.949495          3.286672  
94              0.873583          3.023904  
95              0.771632          2.671002  
96              1.282982          4.854060  
97              1.282800          4.853373  
...                  ...               ...  
1357950         1.598806          0.368673  
1357951         1.619327          0.373405  
1357952         1.707922          0.393834  
1357953         1.679295          0.387233  
1357954         1.268176          0.292432  

[1204408 rows x 16 columns]

In [35]:
df3_combine = pd.merge(df3_actual_return, df3_predict_return, how ='left', on =['Date', 'GVKEY'])
df3_combine

GVKEY       Date    Return  LogSize_-1  LogB/M_-1  Return_-2,-12  \
0          1004 1980-01-31  0.226695    3.340787  -0.056517       0.163113   
1          1004 1980-02-29  0.026087    3.424113  -0.030764       0.275932   
2          1004 1980-03-31 -0.296610    3.548208  -0.056517       0.585662   
3          1004 1980-04-30  0.009639    3.196506   0.295327       0.576707   
4          1004 1980-05-31 -0.012048    3.196647   0.187944       0.222463   
...         ...        ...       ...         ...        ...            ...   
1204403  317264 2020-08-31 -0.011710    6.076221   0.661244      -0.154099   
1204404  317264 2020-09-30 -0.050948    6.063467   0.673023      -0.178057   
1204405  317264 2020-10-31  0.023720    6.010200   0.725315      -0.185330   
1204406  317264 2020-11-30  0.332927    6.027104   0.701871      -0.356627   
1204407  317264 2020-12-31  0.115279    6.307898   0.414494      -0.428572   

         LogIssues_-1,-36  Accruals_Yr-1  ROA_Yr-1  LogAG_Yr-1  DY_-1,-12  \
0                0.436316          2.457  0.039490    0.215465   0.028814   
1                0.545934          2.457  0.039490    0.215465   0.036522   
2                0.645355          2.457  0.039490    0.215465   0.029153   
3                0.646576         -1.103  0.049795    0.127634   0.041446   
4                0.647798         -1.103  0.049795    0.127634   0.051084   
...                   ...            ...       ...         ...        ...   
1204403         -0.077902       -115.174 -0.030312   -0.065911   0.000000   
1204404         -0.078937       -115.174 -0.030312   -0.065911   0.000000   
1204405         -0.079973       -115.174 -0.030312   -0.065911   0.000000   
1204406         -0.086573       -115.174 -0.030312   -0.065911   0.000000   
1204407         -0.093217       -115.174 -0.030312   -0.065911   0.000000   

         LogReturn_-13,-36  LogIssues_-1,-12  Turnover_-1,-12  \
0                 0.338005          0.131147     21231.256206   
1                 0.825340          0.239829     30526.102419   
2                 0.639495          0.337378     32507.391326   
3                 0.797559          0.336727     33503.947702   
4                 0.418066          0.336076     34249.060531   
...                    ...               ...              ...   
1204403           0.270834         -0.071939    252262.069994   
1204404           0.514577         -0.070355    252022.588874   
1204405           0.519062         -0.067314    257688.600552   
1204406           0.741259         -0.069822    252706.685796   
1204407           0.987534         -0.072356    255335.436094   

         Debt/Price_Yr-1  Sales/Price_Yr-1  Forecast/Predicted Return  
0               0.949495          3.286672                   0.013048  
1               0.873583          3.023904                   0.012534  
2               0.771632          2.671002                   0.013856  
3               1.282982          4.854060                   0.016681  
4               1.282800          4.853373                   0.021450  
...                  ...               ...                        ...  
1204403         1.598806          0.368673                   0.006989  
1204404         1.619327          0.373405                   0.007863  
1204405         1.707922          0.393834                   0.006656  
1204406         1.679295          0.387233                   0.005834  
1204407         1.268176          0.292432                   0.007153  

[1204408 rows x 17 columns]

In [36]:
result = pd.DataFrame()
result['Date'] = df3_combine['Date']
result['GVKEY'] = df3_combine['GVKEY']
result['Actual Return'] = df3_combine['Return']
result['Forecast/Predicted Return'] = df3_combine['Forecast/Predicted Return']

result = result.dropna()
# result.set_index('Date', inplace=True)
result.sort_values(by='Date', inplace = True)
result = result.set_index(['Date','GVKEY'])
result

Actual Return  Forecast/Predicted Return
Date       GVKEY                                           
1980-01-31 1004         0.226695                   0.013048
           9514         0.006849                   0.014782
           9507         0.032967                   0.008504
           2263        -0.012422                   0.010768
           2265        -0.032258                   0.014042
...                          ...                        ...
2021-03-31 141459       0.068054                   0.010241
           119275       0.124439                   0.011083
           3905         0.006107                   0.012030
           31567       -0.089674                   0.016385
           120093       0.165741                   0.009318

[1172862 rows x 2 columns]

# Question 4

In [37]:
result['Position'] = np.sign(result['Forecast/Predicted Return'])
result['Strategy'] = result['Position'] * result['Actual Return']
result_updated = result.reset_index()
result_updated['Year_Time'] = result_updated['Date'].dt.year
result_updated

Date   GVKEY  Actual Return  Forecast/Predicted Return  \
0       1980-01-31    1004       0.226695                   0.013048   
1       1980-01-31    9514       0.006849                   0.014782   
2       1980-01-31    9507       0.032967                   0.008504   
3       1980-01-31    2263      -0.012422                   0.010768   
4       1980-01-31    2265      -0.032258                   0.014042   
...            ...     ...            ...                        ...   
1172857 2021-03-31  141459       0.068054                   0.010241   
1172858 2021-03-31  119275       0.124439                   0.011083   
1172859 2021-03-31    3905       0.006107                   0.012030   
1172860 2021-03-31   31567      -0.089674                   0.016385   
1172861 2021-03-31  120093       0.165741                   0.009318   

         Position  Strategy  Year_Time  
0             1.0  0.226695       1980  
1             1.0  0.006849       1980  
2             1.0  0.032967       1980  
3             1.0 -0.012422       1980  
4             1.0 -0.032258       1980  
...           ...       ...        ...  
1172857       1.0  0.068054       2021  
1172858       1.0  0.124439       2021  
1172859       1.0  0.006107       2021  
1172860       1.0 -0.089674       2021  
1172861       1.0  0.165741       2021  

[1172862 rows x 7 columns]

In [38]:
annualized_return_each_year = result_updated.groupby(['Year_Time'])['Strategy'].mean()*12
annualized_return_each_year

Year_Time
1980    0.296278
1981    0.058129
1982    0.298557
1983    0.344448
1984   -0.015679
1985    0.223345
1986    0.097319
1987    0.004721
1988    0.206347
1989    0.149700
1990   -0.133643
1991    0.299800
1992    0.209629
1993    0.225053
1994    0.035094
1995    0.253904
1996    0.188841
1997    0.214043
1998   -0.031244
1999    0.245751
2000    0.071436
2001    0.211647
2002   -0.062930
2003    0.485730
2004    0.230207
2005    0.088130
2006    0.183678
2007    0.019401
2008   -0.472056
2009    0.421363
2010    0.281097
2011   -0.047826
2012    0.159276
2013    0.350198
2014    0.058595
2015   -0.076914
2016    0.195487
2017    0.158085
2018   -0.118998
2019    0.235874
2020    0.272459
2021    0.618604
Name: Strategy, dtype: float64

In [39]:
# overall annually_return
annually_return = result[['Strategy']].mean()*12
annually_return 

Strategy    0.139776
dtype: float64

In [40]:
Accuracy = np.sign(result["Actual Return"] *
                   result["Forecast/Predicted Return"]
                  ).value_counts()
Accuracy_score = (Accuracy.values[0] / sum(Accuracy))
Accuracy_score

0.5066529566138216

# Question 5

### Method 1: 
#### Removing Insignificant Factors

In [41]:
df4 = df3.copy()
formula_4 = 'Return ~ 1 + Q("LogSize_-1") + Q("LogB/M_-1") + Q("Return_-2,-12") + Q("Accruals_Yr-1") + Q("ROA_Yr-1") + Q("LogAG_Yr-1")+ Q("Sales/Price_Yr-1")'
def ols_coef(x,formula):
    return smf.ols(formula,data=x).fit()

reg_4 = df4.groupby('Date').apply(ols_coef, formula_4)
reg_4_list = list(reg_4.values)
model_4 = Regtable(reg_4_list,bfmt='.5f')
model_4 = model_4.render().T
model_4.index = reg_4.index
model_4 = model_4[model_4['obs']>=40]
model_4.drop(columns=[''], inplace =True)
model_4.rename(columns = lambda x: x.strip('Q("")'), inplace = True)
def func(model):
    for i in model.columns:
        model[i] = model[i].apply(lambda x: pd.to_numeric(x))
    return model
model_4 = func(model_4)

moving_average2 = pd.DataFrame()
moving_average2["Intercept_120"] = model_4["Intercept"].rolling(window = 120).mean()
moving_average2["LogSize_-1_120"] = model_4["LogSize_-1"].rolling(window = 120).mean()
moving_average2["LogB/M_-1_120"] = model_4["LogB/M_-1"].rolling(window = 120).mean()
moving_average2["Return_-2,-12_120"] = model_4["Return_-2,-12"].rolling(window = 120).mean()
moving_average2["Accruals_Yr-1_120"] = model_4["Accruals_Yr-1"].rolling(window = 120).mean()
moving_average2["ROA_Yr-1_120"] = model_4["ROA_Yr-1"].rolling(window = 120).mean()
moving_average2["LogAG_Yr-1_120"] = model_4["LogAG_Yr-1"].rolling(window = 120).mean()
moving_average2["Sales/Price_Yr-1_120"] = model_4["Sales/Price_Yr-1"].rolling(window = 120).mean()
moving_average2 = moving_average2.dropna()

start_date = '1979-12-31'
after_start_date = df4["Date"] >= start_date
df4_actual = df4.loc[after_start_date]

df4_predict = pd.merge(df4_actual, moving_average2, how ='left', on ='Date')

df4_predict['Forecast/Predicted Return']\
= df4_predict['LogSize_-1']*df4_predict['LogSize_-1_120'] + df4_predict['LogB/M_-1']*df4_predict['LogB/M_-1_120']\
  + df4_predict['Return_-2,-12']*df4_predict['Return_-2,-12_120']\
  + df4_predict['Accruals_Yr-1']*df4_predict['Accruals_Yr-1_120'] + df4_predict['ROA_Yr-1']*df4_predict['ROA_Yr-1_120']\
  + df4_predict['LogAG_Yr-1']*df4_predict['LogAG_Yr-1_120'] + \
  + df4_predict['Sales/Price_Yr-1']*df4_predict['Sales/Price_Yr-1_120'] + df4_predict['Intercept_120']

df4_predict_return = pd.DataFrame()
df4_predict_return['Date'] = df4_predict['Date'] + pd.offsets.MonthEnd(1)
df4_predict_return['GVKEY'] = df4_predict['GVKEY']
df4_predict_return['Forecast/Predicted Return'] = df4_predict['Forecast/Predicted Return']
df4_predict_return.sort_values(by = 'Date')

start_date2 = '1980-01-31'
after_start_date2 = df4["Date"] >= start_date2
df4_actual_return = df4.loc[after_start_date2]

df4_combine = pd.merge(df4_actual_return, df4_predict_return, how ='left', on =['Date', 'GVKEY'])

result2 = pd.DataFrame()
result2['Date'] = df4_combine['Date']
result2['GVKEY'] = df4_combine['GVKEY']
result2['Actual Return'] = df4_combine['Return']
result2['Forecast/Predicted Return'] = df4_combine['Forecast/Predicted Return']

result2 = result2.dropna()
# result.set_index('Date', inplace=True)
result2.sort_values(by='Date', inplace = True)
result2 = result2.set_index(['Date','GVKEY'])

result2['Position'] = np.sign(result2['Forecast/Predicted Return'])
result2['Strategy'] = result2['Position'] * result2['Actual Return']
result2_updated = result2.reset_index()
result2_updated['Year_Time'] = result2_updated['Date'].dt.year

annualized_return_each_year2 = result2_updated.groupby(['Year_Time'])['Strategy'].mean()*12

annually_return2 = result2[['Strategy']].mean()*12

Accuracy2 = np.sign(result2["Actual Return"] *
                   result2["Forecast/Predicted Return"]
                  ).value_counts()
Accuracy2_score = (Accuracy2.values[0] / sum(Accuracy2))
Accuracy2_score

0.5071824187252618

## Method 2:

#### Increasing Rolling Window + Removing Insignificant Characteristics

In [42]:
df4 = df3.copy()
formula_4 = 'Return ~ 1 + Q("LogSize_-1") + Q("LogB/M_-1") + Q("Return_-2,-12") + Q("Accruals_Yr-1") + Q("ROA_Yr-1") + Q("LogAG_Yr-1")+ Q("Sales/Price_Yr-1")'
def ols_coef(x,formula):
    return smf.ols(formula,data=x).fit()

reg_4 = df4.groupby('Date').apply(ols_coef, formula_4)
reg_4_list = list(reg_4.values)
model_4 = Regtable(reg_4_list,bfmt='.5f')
model_4 = model_4.render().T
model_4.index = reg_4.index
model_4 = model_4[model_4['obs']>=40]
model_4.drop(columns=[''], inplace =True)
model_4.rename(columns = lambda x: x.strip('Q("")'), inplace = True)
def func(model):
    for i in model.columns:
        model[i] = model[i].apply(lambda x: pd.to_numeric(x))
    return model
model_4 = func(model_4)

moving_average2 = pd.DataFrame()
moving_average2["Intercept_240"] = model_4["Intercept"].rolling(window = 240).mean()
moving_average2["LogSize_-1_240"] = model_4["LogSize_-1"].rolling(window = 240).mean()
moving_average2["LogB/M_-1_240"] = model_4["LogB/M_-1"].rolling(window = 240).mean()
moving_average2["Return_-2,-12_240"] = model_4["Return_-2,-12"].rolling(window = 240).mean()
moving_average2["Accruals_Yr-1_240"] = model_4["Accruals_Yr-1"].rolling(window = 240).mean()
moving_average2["ROA_Yr-1_240"] = model_4["ROA_Yr-1"].rolling(window = 240).mean()
moving_average2["LogAG_Yr-1_240"] = model_4["LogAG_Yr-1"].rolling(window = 240).mean()
moving_average2["Sales/Price_Yr-1_240"] = model_4["Sales/Price_Yr-1"].rolling(window = 240).mean()
moving_average2 = moving_average2.dropna()

start_date = '1979-12-31'
after_start_date = df4["Date"] >= start_date
df4_actual = df4.loc[after_start_date]

df4_predict = pd.merge(df4_actual, moving_average2, how ='left', on ='Date')

df4_predict['Forecast/Predicted Return']\
= df4_predict['LogSize_-1']*df4_predict['LogSize_-1_240'] + df4_predict['LogB/M_-1']*df4_predict['LogB/M_-1_240']\
  + df4_predict['Return_-2,-12']*df4_predict['Return_-2,-12_240']\
  + df4_predict['Accruals_Yr-1']*df4_predict['Accruals_Yr-1_240'] + df4_predict['ROA_Yr-1']*df4_predict['ROA_Yr-1_240']\
  + df4_predict['LogAG_Yr-1']*df4_predict['LogAG_Yr-1_240'] + \
  + df4_predict['Sales/Price_Yr-1']*df4_predict['Sales/Price_Yr-1_240'] + df4_predict['Intercept_240']

df4_predict_return = pd.DataFrame()
df4_predict_return['Date'] = df4_predict['Date'] + pd.offsets.MonthEnd(1)
df4_predict_return['GVKEY'] = df4_predict['GVKEY']
df4_predict_return['Forecast/Predicted Return'] = df4_predict['Forecast/Predicted Return']
df4_predict_return.sort_values(by = 'Date')

start_date2 = '1980-01-31'
after_start_date2 = df4["Date"] >= start_date2
df4_actual_return = df4.loc[after_start_date2]

df4_combine = pd.merge(df4_actual_return, df4_predict_return, how ='left', on =['Date', 'GVKEY'])

result2 = pd.DataFrame()
result2['Date'] = df4_combine['Date']
result2['GVKEY'] = df4_combine['GVKEY']
result2['Actual Return'] = df4_combine['Return']
result2['Forecast/Predicted Return'] = df4_combine['Forecast/Predicted Return']

result2 = result2.dropna()
# result.set_index('Date', inplace=True)
result2.sort_values(by='Date', inplace = True)
result2 = result2.set_index(['Date','GVKEY'])

result2['Position'] = np.sign(result2['Forecast/Predicted Return'])
result2['Strategy'] = result2['Position'] * result2['Actual Return']
result2_updated = result2.reset_index()
result2_updated['Year_Time'] = result2_updated['Date'].dt.year

annualized_return_each_year2 = result2_updated.groupby(['Year_Time'])['Strategy'].mean()*12

annually_return2 = result2[['Strategy']].mean()*12

Accuracy2 = np.sign(result2["Actual Return"] *
                   result2["Forecast/Predicted Return"]
                  ).value_counts()
Accuracy2_score = (Accuracy2.values[0] / sum(Accuracy2))
Accuracy2_score

0.5091813119841921

## Method 3:
#### Changing Start Date(Sub-Period) + Increasing Rolling Window + Removing Insignificant Characteristic

In [43]:
df4 = df3.copy()
formula_4 = 'Return ~ 1 + Q("LogSize_-1") + Q("LogB/M_-1") + Q("Return_-2,-12") + Q("Accruals_Yr-1") + Q("ROA_Yr-1") + Q("LogAG_Yr-1")+ Q("Sales/Price_Yr-1")'
def ols_coef(x,formula):
    return smf.ols(formula,data=x).fit()

reg_4 = df4.groupby('Date').apply(ols_coef, formula_4)
reg_4_list = list(reg_4.values)
model_4 = Regtable(reg_4_list,bfmt='.5f')
model_4 = model_4.render().T
model_4.index = reg_4.index
model_4 = model_4[model_4['obs']>=40]
model_4.drop(columns=[''], inplace =True)
model_4.rename(columns = lambda x: x.strip('Q("")'), inplace = True)
def func(model):
    for i in model.columns:
        model[i] = model[i].apply(lambda x: pd.to_numeric(x))
    return model
model_4 = func(model_4)

moving_average2 = pd.DataFrame()
moving_average2["Intercept_240"] = model_4["Intercept"].rolling(window = 240).mean()
moving_average2["LogSize_-1_240"] = model_4["LogSize_-1"].rolling(window = 240).mean()
moving_average2["LogB/M_-1_240"] = model_4["LogB/M_-1"].rolling(window = 240).mean()
moving_average2["Return_-2,-12_240"] = model_4["Return_-2,-12"].rolling(window = 240).mean()
moving_average2["Accruals_Yr-1_240"] = model_4["Accruals_Yr-1"].rolling(window = 240).mean()
moving_average2["ROA_Yr-1_240"] = model_4["ROA_Yr-1"].rolling(window = 240).mean()
moving_average2["LogAG_Yr-1_240"] = model_4["LogAG_Yr-1"].rolling(window = 240).mean()
moving_average2["Sales/Price_Yr-1_240"] = model_4["Sales/Price_Yr-1"].rolling(window = 240).mean()
moving_average2 = moving_average2.dropna()

start_date = '2009-12-31'
after_start_date = df4["Date"] >= start_date
df4_actual = df4.loc[after_start_date]

df4_predict = pd.merge(df4_actual, moving_average2, how ='left', on ='Date')

df4_predict['Forecast/Predicted Return']\
= df4_predict['LogSize_-1']*df4_predict['LogSize_-1_240'] + df4_predict['LogB/M_-1']*df4_predict['LogB/M_-1_240']\
  + df4_predict['Return_-2,-12']*df4_predict['Return_-2,-12_240']\
  + df4_predict['Accruals_Yr-1']*df4_predict['Accruals_Yr-1_240'] + df4_predict['ROA_Yr-1']*df4_predict['ROA_Yr-1_240']\
  + df4_predict['LogAG_Yr-1']*df4_predict['LogAG_Yr-1_240'] + \
  + df4_predict['Sales/Price_Yr-1']*df4_predict['Sales/Price_Yr-1_240'] + df4_predict['Intercept_240']

df4_predict_return = pd.DataFrame()
df4_predict_return['Date'] = df4_predict['Date'] + pd.offsets.MonthEnd(1)
df4_predict_return['GVKEY'] = df4_predict['GVKEY']
df4_predict_return['Forecast/Predicted Return'] = df4_predict['Forecast/Predicted Return']
df4_predict_return.sort_values(by = 'Date')

start_date2 = '2010-01-31'
after_start_date2 = df4["Date"] >= start_date2
df4_actual_return = df4.loc[after_start_date2]

df4_combine = pd.merge(df4_actual_return, df4_predict_return, how ='left', on =['Date', 'GVKEY'])

result2 = pd.DataFrame()
result2['Date'] = df4_combine['Date']
result2['GVKEY'] = df4_combine['GVKEY']
result2['Actual Return'] = df4_combine['Return']
result2['Forecast/Predicted Return'] = df4_combine['Forecast/Predicted Return']

result2 = result2.dropna()
# result.set_index('Date', inplace=True)
result2.sort_values(by='Date', inplace = True)
result2 = result2.set_index(['Date','GVKEY'])

result2['Position'] = np.sign(result2['Forecast/Predicted Return'])
result2['Strategy'] = result2['Position'] * result2['Actual Return']
result2_updated = result2.reset_index()
result2_updated['Year_Time'] = result2_updated['Date'].dt.year

annualized_return_each_year2 = result2_updated.groupby(['Year_Time'])['Strategy'].mean()*12

annually_return2 = result2[['Strategy']].mean()*12

Accuracy2 = np.sign(result2["Actual Return"] *
                   result2["Forecast/Predicted Return"]
                  ).value_counts()
Accuracy2_score = (Accuracy2.values[0] / sum(Accuracy2))
Accuracy2_score

0.5293459261152834

# Conclusion:

> After we tried three methods, the third method gave us the highest accuracy score, 52.93% which is around 2% higher than our original forecast prediction accuray,50.66%.